eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs

```diff
@@ -160,43 +160,81 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+--   (e) is the size of the registered DReps, lookup
+--   (f) is the size of the PoolDistr, insert
 computeDRepDistr ::
   forall c.
   Map (Credential 'Staking c) (CompactForm Coin) ->
   Map (Credential 'DRepRole c) (DRepState c) ->
   Map (Credential 'Staking c) (CompactForm Coin) ->
+  PoolDistr c ->
   Map (DRep c) (CompactForm Coin) ->
   Map (Credential 'Staking c) (UMElem c) ->
-  Map (DRep c) (CompactForm Coin)
-computeDRepDistr stakeDistr regDReps proposalDeposits dRepDistr uMapChunk =
-  Map.foldlWithKey' go dRepDistr uMapChunk
+  (Map (DRep c) (CompactForm Coin), PoolDistr c)
+computeDRepDistr stakeDistr regDReps proposalDeposits poolDistr dRepDistr uMapChunk =
+  Map.foldlWithKey' go (dRepDistr, poolDistr) uMapChunk
   where
-    go accum stakeCred umElem =
+    go (drepAccum, poolAccum) stakeCred umElem =
       let stake = fromMaybe (CompactCoin 0) $ Map.lookup stakeCred stakeDistr
           proposalDeposit = fromMaybe (CompactCoin 0) $ Map.lookup stakeCred proposalDeposits
           stakeAndDeposits = addCompact stake proposalDeposit
-       in case umElemDRepDelegatedReward umElem of
-            Just (r, drep@DRepAlwaysAbstain) ->
-              Map.insertWith addCompact drep (addCompact stakeAndDeposits r) accum
-            Just (r, drep@DRepAlwaysNoConfidence) ->
-              Map.insertWith addCompact drep (addCompact stakeAndDeposits r) accum
-            Just (r, drep@(DRepCredential drepCred))
-              | Map.member drepCred regDReps ->
-                  Map.insertWith addCompact drep (addCompact stakeAndDeposits r) accum
-              | otherwise -> accum
-            Nothing -> accum
+       in case umElemDelegations umElem of
+            Nothing -> (drepAccum, poolAccum)
+            Just (RewardDelegationSPO spo _r) ->
+              ( drepAccum
+              , addToPoolDistr spo proposalDeposit poolAccum
+              )
+            Just (RewardDelegationDRep drep r) ->
+              ( addToDRepDistr drep (addCompact stakeAndDeposits r) drepAccum
+              , poolAccum
+              )
+            Just (RewardDelegationBoth spo drep r) ->
+              ( addToDRepDistr drep (addCompact stakeAndDeposits r) drepAccum
+              , addToPoolDistr spo proposalDeposit poolAccum
+              )
+    addToPoolDistr spo ccoin distr =
+      -- Avoid adding the proposal deposits to anything other than the numerator
+      -- and denominator in order not to interfere with rewards calculation,
+      -- although this is an isolated copy inside the DRepPulser, we want to
+      -- just be sure we do not confuse ourselves over which representations are
+      -- used where.
+      case Map.lookup spo $ distr ^. poolDistrDistrL of
+        Nothing -> distr -- Impossible! Because a delegation to an SPO would also appear in the PoolDistr.
+        Just ips ->
+          distr
+            & poolDistrDistrL %~ Map.insert spo (ips & individualPoolStakeCoinL <>~ ccoin)
+            & poolDistrTotalL <>~ ccoin
```

Comment:
We don't need to lookup or change the SPO distribution if the deposit is zero, which is going to be a 99% of the time.
```suggestion
    addToPoolDistr spo proposalDeposit distr = fromMaybe distr $ do
      guard (proposalDeposit /= mempty)
      ips <- Map.lookup spo $ distr ^. poolDistrDistrL
      pure (distr
            & poolDistrDistrL %~ Map.insert spo (ips & individualPoolStakeCoinL <>~ proposalDeposit)
            & poolDistrTotalL <>~ proposalDeposit
```

Reply:
I thought that the comment would be helpful in making clear that even in the copy of `PoolDistr` that is part of the `DRepPulser` we do not adjust `individualPoolStake :: Rational` with proposal deposits. One can rely on that field being the same between both copies: the one in `NewEpochState` and the one in `DRepPulser`. What do you think? Is this wrong? Should that field be updated?

Reply:
This comment is misleading, because not only DRep pulser has nothing to do with rewards calculation, but also it is totally isolated and cannot interfere with anything except governance, i.e. RATIFY and ENACT rules.

> Should that field be updated?

Nope, leave it unchanged. There is no need to update something that is not used